File: Exercism/python/protein-translation/protein_translation.py
import re


CONDOS_RE_MAP = {
    re.compile("AUG"): "Methionine",
    re.compile("UU[UC]"): "Phenylalanine",
    re.compile("UU[AG]"): "Leucine",
    re.compile("UC[UCAG]"): "Serine",
    re.compile("UA[UC]"): "Tyrosine",
    re.compile("UG[UC]"): "Cysteine",
    re.compile("UGG"): "Tryptophan",
    re.compile("U(?:A[AG]|GA)"): "STOP",
}


def _match_codon(codon):
    for r, match in CONDOS_RE_MAP.items():
        if re.match(r, codon):
            return match
    return None


def _match_codons(strand):
    for i in range(0, len(strand) - 3 + 1, 3):
        match = _match_codon(strand[i:i+3])
        if match == "STOP":
            return
        yield match


def proteins(strand):
    return list(_match_codons(strand))

File: Exercism/haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

import Data.List.Split (chunksOf)
import Data.Map (Map)
import qualified Data.Map as Map
import Text.Read (readMaybe)

data Codon = AUG | UUU | UUC | UUA | UUG | UCU | UCC | UCA | UCG | UAU | UAC | UGU | UGC
           | UGG | UAA  | UAG | UGA
           deriving (Eq, Ord, Show, Read)

data Protein = Methionine
             | Phenylalanine
             | Leucine
             | Serine
             | Tyrosine
             | Cysteine
             | Tryptophan
             | Stop
             deriving (Eq, Show, Read)

proteinCodons :: [(Protein, [Codon])]
proteinCodons = [(Methionine, [AUG])
                ,(Phenylalanine, [UUU, UUC])
                ,(Leucine, [UUA, UUG])
                ,(Serine, [UCU, UCC, UCA, UCG])
                ,(Tyrosine, [UAU, UAC])
                ,(Cysteine, [UGU, UGC])
                ,(Tryptophan, [UGG])
                ,(Stop, [UAA, UAG, UGA])]

proteins :: String -> Maybe [String]
proteins xs = map show . takeWhile (/= Stop) <$> proteinsMaybe xs

proteinsMaybe :: String -> Maybe [Protein]
proteinsMaybe xs = mapM (>>= codonToProtein) (parseCodons xs)

parseCodons :: String -> [Maybe Codon]
parseCodons = map readMaybe . chunksOf 3

codonToProtein :: Codon -> Maybe Protein
codonToProtein = flip Map.lookup codonsToProteinsMap

codonsToProteinsMap :: Map Codon Protein
codonsToProteinsMap = Map.fromList [(y,x) | (x,ys) <- proteinCodons, y <- ys]
